### ISSUE TITLE
Locks overwriting each other and acquired by more than one tab at same time

### ISSUE THREAD

**Issue by brianneisler** [bitovi/web-locks-polyfill#5]
### Problem

Locks are being acquired by more than one tab at the same time when they should not be.

### Reproduction
In my testing of this polyfill package, I am opening three tabs very close together. Occasionally, the lock appears to be acquired by more than one tab at the same time.  Here is my log output from the three separate tabs.

Tab 1
```
[Log] 1616447907354 – "clientId" – "8679140820985151" (0.chunk.js, line 188498)
[Log] useAnonymousSignup - requesting lock - now: – 1616447908142 (main.chunk.js, line 28892)
[Log] 1616447908143 – "8679140820985151:1" – "set innerKey. InnerKey was" – null (0.chunk.js, line 188498)
[Log] 1616447908254 – "8679140820985151:1" – "locked" (0.chunk.js, line 188498)
[Log] useAnonymousSignup - lock acquired - now: – 1616447908255 – " cancelled:" – false – " accessToken:" – undefined (main.chunk.js, line 28894)
[Log] useAnonymousSignup - lock releasing - now: – 1616447909542 (main.chunk.js, line 28904)
[Log] 1616447909551 – "8679140820985151:1" – "unlocked" (0.chunk.js, line 188498)
```

Tab 2
```
[Log] 1616447907700 – "clientId" – "4030941946461747" (0.chunk.js, line 188498)
[Log] useAnonymousSignup - requesting lock - now: – 1616447908380 (main.chunk.js, line 28892)
[Log] 1616447908381 – "4030941946461747:1" – "set innerKey. InnerKey was" – null (0.chunk.js, line 188498)
[Log] 1616447908499 – "4030941946461747:1" – "outerKeyAlreadyOwned. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447908549 – "4030941946461747:1" – "innerKeyAlreadyOwned. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447908570 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447908610 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447909479 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910063 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910146 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910188 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910230 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910353 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910374 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910394 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910570 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910674 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447910828 – "4030941946461747:1" – "innerKeyAlreadyOwned backing off. Owned by" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447911479 – "4030941946461747:1" – "set innerKey. InnerKey was" – null (0.chunk.js, line 188498)
[Log] 1616447912498 – "4030941946461747:1" – "locked" (0.chunk.js, line 188498)
[Log] useAnonymousSignup - lock acquired - now: – 1616447912502 (main.chunk.js, line 28894)
[Log] useAnonymousSignup - lock releasing (main.chunk.js, line 28897)
[Log] 1616447912502 – "4030941946461747:1" – "unlocked" (0.chunk.js, line 188498)
```

Tab 3
``` 
[Log] 1616447907697 – "clientId" – "08631345487356656" (0.chunk.js, line 188498)
useAnonymousSignup - requesting lock - now: – 1616447908390
[Log] 1616447908390 – "08631345487356656:1" – "set innerKey. InnerKey was" – null (0.chunk.js, line 188498)
[Log] 1616447908555 – "08631345487356656:1" – "outerKeyAlreadyOwned. Owned by" – "4030941946461747:1" (0.chunk.js, line 188498)
[Log] 1616447908588 – "08631345487356656:1" – "set innerKey. InnerKey was" – "08631345487356656:1" (0.chunk.js, line 188498)
[Log] 1616447908623 – "08631345487356656:1" – "locked" (0.chunk.js, line 188498)
[Log] useAnonymousSignup - lock acquired - now: – 1616447908623" (main.chunk.js, line 28894)
[Log] useAnonymousSignup - lock releasing - now: – 1616447911212 (main.chunk.js, line 28904)
[Log] 1616447911219 – "08631345487356656:1" – "unlocked" (0.chunk.js, line 188498)
```

Expected Behavior
Lock should only be able to be acquired by one tab at a time.

Environment
- mac OSX 10.15.7
- Safari 14.0.2

**Comment by justinbmeyer** [comment]
Thanks for reporting. Does this only happen in safari for you? Does it happen often?

**Comment by brianneisler** [comment]
Hey @justinbmeyer, thanks for getting back to me! This does not happen on Chrome, but the polyfill is not used there since chrome has the native web locks API available. So that seems to indicate that it's an issue in the polyfill library and not with the application code.

I'm able to reproduce this issue in roughly 1 in 3 attempts, so definitely happening pretty frequently.

**Comment by brianneisler** [comment]
Looking into this a bit further, I found that the same issue is reproducible in Firefox with slightly different behavior. 

This post has good insight into what the underlying issue is...

https://www.javaer101.com/en/article/35579652.html

TLDR: There's no underlying mutex for the storage API across tabs in Firefox (I'm guessing the same might be true for Safari). Instead, each tab keeps a local copy of the localstorage and updates its own copy by listening for the `storage` event. When tab A and tab B write to the same local storage key at the same time they each get notified of each other's storage events. So tab A ends up with tab B's copy and tab B ends up with tab A's. 

You can see this behavior in these logs from Firefox...

Tab A
```
getItem - key: _MUT_IN_LOCK_bubbles-signup-lock  item: {"expiresAt":1616529923384,"value":"754205658451178:1"} make-locks-request.js:81
1616529894916 4715247356955956:1 innerKeyAlreadyOwned. Owned by 754205658451178:1 make-locks-request.js:343
getItem - key: _MUT_IN_LOCK_bubbles-signup-lock  item: {"expiresAt":1616529923384,"value":"754205658451178:1"} make-locks-request.js:81
1616529895920 4715247356955956:1 innerKeyAlreadyOwned backing off. Owned by 754205658451178:1 make-locks-request.js:343
getItem - key: _MUT_IN_LOCK_bubbles-signup-lock  item: {"expiresAt":1616529923384,"value":"754205658451178:1"} make-locks-request.js:81
1616529896921 4715247356955956:1 innerKeyAlreadyOwned backing off. Owned by 754205658451178:1
```

Tab B
```
getItem - key: _MUT_IN_LOCK_bubbles-signup-lock  item: {"expiresAt":1616529923287,"value":"13501602792080392:1"} make-locks-request.js:81
1616529895015 754205658451178:1 innerKeyAlreadyOwned. Owned by 13501602792080392:1 make-locks-request.js:343
getItem - key: _MUT_IN_LOCK_bubbles-signup-lock  item: {"expiresAt":1616529923287,"value":"13501602792080392:1"} make-locks-request.js:81
1616529896016 754205658451178:1 innerKeyAlreadyOwned backing off. Owned by 13501602792080392:1 make-locks-request.js:343
getItem - key: _MUT_IN_LOCK_bubbles-signup-lock  item: {"expiresAt":1616529923287,"value":"13501602792080392:1"} make-locks-request.js:81
1616529897021 754205658451178:1 innerKeyAlreadyOwned backing off. Owned by 13501602792080392:1 make-locks-request.js:343
```

**Comment by brianneisler** [comment]
Ok, so in the above situation for Firefox, when a deadlock like the above occurs, the `lockTimeout` will naturally resolve the deadlock (even though i don't think that was the intended behavior, but i could be wrong). This results in the lock hanging for the duration of the `lockTimeout` so it's a word of caution against raising that value too high. 

In the Safari case, what i believe is happening is that the `tick` is set too low. I'm seeing the same deadlock behavior in Safari after raising the tick which also resolves itself using the `lockTimeout` which indicates that the local storage behavior is similar to what is described in the previous post.

I believe what I was originally seeing in Safari was a result of the next tick occurring before the localStorage update events had a chance to be dispatched.

**Comment by justinbmeyer** [comment]
@brianneisler .... sorry about not responding, I somehow missed this response (!!!).

I ran this over and over again in FF and safari and never saw this issue.  I'll try it again this weekend.

Thanks for the info on the events.  I'm trying to think of a way to improve this w/o setting tick higher.  What did you set it at?

I also thought deadlocks would "naturally" be resolved by the random back-off.  It's odd that `lockTimeout` needs to be used.

Thanks again for reporting this.  I want to figure out "some weird trick" to make it work cross browser.

Btw, what are you using this for?  My use case is to prevent refresh_tokens from being used at the same time across multiple tabs using the same auth tokens.

**Comment by brianneisler** [comment]
hey @justinbmeyer, thanks for getting back to me. 

Our use case at usebubbles.com is the same. We're trying to prevent refresh_tokens from being used at the same time when multiple tabs are opened up very quickly. We're using revolving refresh tokens so simultaneous calls actually invalidate a user's entire refresh token chain which is a major issue. 

Here was the config i ended up using that seemed to work pretty well
```
lockTimeout: 2000, 
tick: 100 
```

It would definitely be great if this could work flawlessly in multiple browsers. 

>I also thought deadlocks would "naturally" be resolved by the random back-off. It's odd that lockTimeout needs to be used.

The reason they're not resolved by the random back-off is that each tab thinks the other owns the inner lock. There's no resolution for that in the code other than to wait for the lockTimeout to expire.

https://github.com/bitovi/web-locks-polyfill/blob/1fd83c99bd24f4638d5e0869db2d32f6c5007361/make-locks-request.js#L166-L175

https://github.com/bitovi/web-locks-polyfill/blob/1fd83c99bd24f4638d5e0869db2d32f6c5007361/make-locks-request.js#L236-L237

You can see that the above state will just continually repeat on both sides since there's no escape once a lock request thinks someone else owns the inner-lock. I think realizing that the local storage is not backed by a mutex in Safari and Firefox is the heart of the issue and that you can end up in these states where each tab has a different idea of what is contained in local storage.

One thought I had that might ease the lock tension is if the lock requests that are in the `"requested"` state first checked the outer lock using `getItem` instead of forcefully setting the outer lock in that situation. This would give the owner of the inner lock that is in the state `"set-outer"` a chance to set the outer lock without being interrupted. This wouldn't resolve the above issue but it would help the locks resolve faster and potentially allow the tick to remain low.

https://github.com/bitovi/web-locks-polyfill/blob/1fd83c99bd24f4638d5e0869db2d32f6c5007361/make-locks-request.js#L166-L169

Anyways, let me know if the above makes sense. Happy to open up a PR. It would also be good to write up a test that exercises these possible misalignment states.

**Comment by brianneisler** [comment]
hey @justinbmeyer, just raising this one again to see if there's been any progress. Thanks!

**Comment by aermin** [comment]
> hey @justinbmeyer, just raising this one again to see if there's been any progress. Thanks!

Hi @brianneisler, maybe you could try this one https://github.com/aermin/web-locks. Hope it could help you.

**Comment by justinbmeyer** [comment]
Yeah, I've been swamped by other things.  If that other one works, please let me know.  Thanks!
